nonuniform tensor parallelism

daiyaanarfeen opened the pull request:
# What does this PR do ?
Enables training with reduced TP degree on DP ranks with some failed GPUs. Failed ranks can be specified by their coordinate in the parallelism mapping. This can enable training jobs to continue without rescheduling and waiting for available resources. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 1): Add PR label `Expert Review`

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

skyw commented on megatron/core/distributed/nonuniform_tp.py:
```diff
@@ -0,0 +1,404 @@
+        ntp_map(layer.mlp, ddp_config, layer.mlp.config.ffn_hidden_size)
+
+
+def test_ntp():
```
move test to the right place under /tests.

skyw commented on megatron/core/parallel_state.py:
```diff
@@ -903,62 +972,173 @@ def initialize_model_parallel(
-        if rank in ranks:
-            _MODEL_PARALLEL_GROUP = group
-            _MODEL_PARALLEL_GLOBAL_RANKS = ranks
+                _CONTEXT_PARALLEL_GROUP = group
```
update to latest, looks like you branched from very old code?

skyw reviewed (commented):
Recommend to put everything into nununiform_tp.py, inherit from core classes and override member function when needed. That way code can be non-intrusive.

daiyaanarfeen committed:
feat: non-intrusive nonuniform tensor parallelism implementation

- All NTP logic contained in nonuniform_tp.py as subclasses
- NonuniformTPDistributedDataParallel: inherits from DistributedDataParallel
- NonuniformTPParamAndGradBuffer: handles gradient buffer splitting for NTP
- NonuniformTPOptimizer: wrapper for gradient contiguity
- initialize_nonuniform_tp_process_groups(): reconfigures process groups after init
- Only config changes to core files (distributed_data_parallel_config.py)
- Added comprehensive CLAUDE.md documentation

daiyaanarfeen committed:
test: add comprehensive unit tests for nonuniform TP

- Moved test from nonuniform_tp.py to tests/unit_tests/distributed/
- Added TestNonuniformTPUtilities: tests for utility functions
  - compute_uniform_tp_spares_with_parity (3 test cases)
  - get_active_ranks_for_dp (2 test cases)
- Added TestNonuniformTPParameterResharding: tests for parameter resharding
  - ntp_map for no spares, healthy ranks, unhealthy ranks
  - ntp_init for layers with attention and MLP (4 test cases)
- Added TestNonuniformTPOptimizer: tests for optimizer wrapper
  - attribute delegation, prepare_grads, contiguity handling (5 test cases)
- Added TestNonuniformTPIntegration: integration tests
  - DDP initialization and backward hooks (2 test cases)
- Total: 17 test cases covering all major NTP functionality

daiyaanarfeen committed:
fix: remove read-only shape attribute assignment in test

PyTorch tensors have shape as a read-only property, no need to set it

daiyaanarfeen committed:
fix: remove another shape assignment in test_ntp_backward_hook_core_gpu

daiyaanarfeen committed:
test: add end-to-end NTP test for 8 GPUs without mocking

- Tests 2 DP workers: DP rank 0 with TP=2 (reduced), DP rank 1 with TP=4 (healthy)
- Uses tp_base=4, tp_spares=2 configuration
- Verifies process group reconfiguration
- Tests parameter initialization and gradient computation
- No mocking - actual distributed test with real model

daiyaanarfeen committed:
fix: skip spare ranks gracefully in end-to-end test

Spare ranks would call sys.exit(0) during NTP initialization,
which pytest treats as a failure. Now spare ranks skip the test
gracefully before that happens.

daiyaanarfeen committed:
refactor: move NTP config to separate NonuniformTPConfig class

- Create NonuniformTPConfig dataclass in nonuniform_tp.py
- Remove NTP fields from DistributedDataParallelConfig (non-intrusive)
- Update all NTP functions/classes to use NonuniformTPConfig
- Update all tests to use NonuniformTPConfig
- Update CLAUDE.md documentation

This makes the NTP implementation completely self-contained with
zero modifications to core Megatron files.

daiyaanarfeen committed:
chore: remove unused typing imports from distributed_data_parallel_config.py

daiyaanarfeen committed:
test: fix integration tests to actually run without distributed mock

Replace always-skipping integration tests with ones that use a real
single-GPU distributed context (tp_base=1, tp_spares=0), so they
run and pass in single-process mode via Utils.initialize_model_parallel.

plugyawn commented:
@daiyaanarfeen @skyw looks like a good feature to have for core, could lend a hand in tidying this up.

daiyaanarfeen commented:
@plugyawn the feature has been merged to dev branch, you can check it out here: https://github.com/NVIDIA/Megatron-LM/pull/4585
